### DISCUSSION TITLE
How to implement vjournal?

### DISCUSSION THREAD

**Discussion by Terramoto** [TechbeeAT/jtxBoard#681] [General]
If i wanted to build an app or a extension for something like Thunderbird how does one go about to implement vjournal?

**Comment by rogercreagh** [comment]
Find a calDav library that works for whatever language/dev environment you use. i'm aware of php and python ones, I imagine there are others - obviously there must be something for android that Patrick uses - google around.I've not found one with native vjournal awareness though so you might end up adapting it.

The tricky bits are communicating with the server (jtxB uses Davx5 to do that) - and then parsing the data you get back - which involves reading and understanding all of the relevant rfc's so you know what to expect. 

If you are using an existing app or library to talk with the server, you'll possibly find yourself writing the vjournal specific bits yourself based on how it handles vevents.

Oh and designing a suitable data structure to ensure you don't loose anything important for other apps because you are not using a particular property or parameter. 

Plus of course the UI and deciding how you'll handle vjournals without  start date - jtxB has led the way in treating those as a separate type of object called a note -there isn't a vnote component as such, but you'll probably want to treat them differently from the users point of view as jtxB does - very useful. 

Good luck, I'm making slow progress with a web site component in php - slow not cos its particularly difficult, but because I don't have a lot of time for it, especially in spring summer and autumn.

**Reply by 855579762735** [reply]
Hey,
I like to selfhost things and I've been on a hunt for a good web interface counterpart to my radicale and jtxboard setup. I would be interested in this web interface you're designing in php. Is it an open source repository ? I'd be interested in taking a peak. I like to call myself a DAVocate just because of how much I enjoy the protocol and how well everything works together, but I've struggled to fine a desktop / webui that works well with VJournal ( or at least as well as jtxboard does ). Thanks!

**Reply by rogercreagh** [reply]
Its a Joomla component, so assumes a Joomla website. Its a long way from usable yet - all it does so far is read Jounral entries from the CalDAV server (Nextcloud) and store the data in the local database (joomla). I doubt I'll be doing anything more on it before next winter. Its quite handy for me being able to read and search on a proper screen. I write on a tablet.

**Reply by 855579762735** [reply]
Regarding desktop applications .. how difficult might it be to develop a plugin for something like LogSeq that retrieves the markdown files and syncs then to VJournal ? One would think that would no be too hard but I am not sure where to start regarding the development of such a plugin. The LogSeq app doesn't even support basic sync over WebDAV so maybe the API is just missing a sync component, not quite sure. I'd be open to suggestions for other Desktop apps that might be capable to syncing this way.

**Comment by patrickunterwegs** [comment]
Hi @Terramoto !

The base anyhow is https://datatracker.ietf.org/doc/html/rfc5545 for the data structure. As @rogercreagh points out, it is important to think of how to store the data locally. 

If you're thinking about doing something for Thunderbird specifically, I'd say you can reuse a lot of existing code, I'm not sure if this can be done through an extension though. Thunderbird already supports the VTODO component and comes with all necessary functionality to store a CalDAV-Account and sync the entries. There shouldn't be any restriction if you sync an entry with a VTODO or a VJOURNAL as long as the CalDAV-Server itself supports it. 

So Thunderbird already deals with VTODOs (additionally to VEVENTs). The good thing is that VJOURNALs use only a subset of the fields used in VTODOs. There would be nothing new to consider, just to see what's there and adapt it to also support VJOURNALs. 
For example: While VTODOs have a DTSTART, DUE and COMPLETED property, the VJOURNAL only has the DTSTART property to define a date/time of the entry. Another example: While VTODOs can have a VALARM, this is not defined for VJOURNALs.

I assume the challenge would be more to get through the code of Thunderbird and to make the right adaptions. I also think that the VJOURNAL should be implemented within Thunderbird (and not through an extension), so I assume there's also some alignment necessary with the project owners.

If you're thinking about something like this, watch out for programs that give financial support for open source projects 😉

**Comment by bepolymathe** [comment]
Thanks @Terramoto ! It would be a wonderful addition to the ecosystem. Is it still in the pipeline? If you need some beta tester, ask me...

**Comment by bvdbos** [comment]
About 17 years ago [I was hoping to get the TB-devs to create a Vjournal-extension](https://bugzilla.mozilla.org/show_bug.cgi?id=185537#c15) but no luck... But with the help of chatgpt there's a [caldav-bridge for Nextcloud Notes](https://apps.nextcloud.com/apps/notes_caldav) and for [JournalNotes](https://apps.nextcloud.com/apps/journalnotes) I also vibecoded [a bridge](https://apps.nextcloud.com/apps/journal_caldav).
